stl/inc/ranges

```diff
@@ -2635,6 +2589,421 @@ namespace ranges {
+    // clang-format off
+    template <input_range _Vw>
+        requires view<_Vw> && input_range<range_reference_t<_Vw>>
+            && (is_reference_v<range_reference_t<_Vw>> || view<range_value_t<_Vw>>)
+    class join_view;
+    // clang-format on
+
+    template <class _Vw>
+    class _Join_view_base : public view_interface<join_view<_Vw>> {
+    protected:
+        /* [[no_unique_address]] */ views::all_t<range_reference_t<_Vw>> _Inner{};
+    };
+
+    // clang-format off
+    template <class _Vw>
+        requires is_reference_v<range_reference_t<_Vw>>
+    class _Join_view_base<_Vw> : public view_interface<join_view<_Vw>> {};
+
+    template <input_range _Vw>
+        requires view<_Vw> && input_range<range_reference_t<_Vw>>
+            && (is_reference_v<range_reference_t<_Vw>> || view<range_value_t<_Vw>>)
+    class join_view : public _Join_view_base<_Vw> {
+        // clang-format on
+    private:
+        using _InnerRng = range_reference_t<_Vw>;
+        /* [[no_unique_address]] */ _Vw _Range{};
+
+        template <bool _Const>
+        class _Sentinel;
+
+        template <class _OuterTraits, class _InnerTraits, bool _Deref_is_glvalue,
+            bool _Inner_common> // TRANSITION, LWG-3289
+        struct _Category_base {};
+
+        // clang-format off
+        template <class _OuterTraits, class _InnerTraits, bool _Deref_is_glvalue, bool _Inner_common>
+            requires _Has_member_iterator_category<_OuterTraits> && _Has_member_iterator_category<_InnerTraits>
+        struct _Category_base<_OuterTraits, _InnerTraits, _Deref_is_glvalue, _Inner_common> {
+            using iterator_category =
+                conditional_t<_Deref_is_glvalue && _Inner_common // per LWG issue unnumbered as of 2021-03-16
+                           && derived_from<typename _OuterTraits::iterator_category, bidirectional_iterator_tag>
+                           && derived_from<typename _InnerTraits::iterator_category, bidirectional_iterator_tag>,
+                    bidirectional_iterator_tag,
+                    conditional_t<_Deref_is_glvalue
+                               && derived_from<typename _OuterTraits::iterator_category, forward_iterator_tag>
+                               && derived_from<typename _InnerTraits::iterator_category, forward_iterator_tag>,
+                        forward_iterator_tag,
+                        conditional_t<derived_from<typename _OuterTraits::iterator_category, input_iterator_tag>
+                                   && derived_from<typename _InnerTraits::iterator_category, input_iterator_tag>,
+                            input_iterator_tag,
+                            output_iterator_tag>>>;
+        };
+        // clang-format on
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<iterator_traits<iterator_t<_Maybe_const<_Const, _Vw>>>,
+                              iterator_traits<iterator_t<range_reference_t<_Maybe_const<_Const, _Vw>>>>,

```

Comment:
Maybe we should also give `range_reference_t<_Maybe_const<_Const, _Vw>>` a name?

Would it make sense to do 
```cpp
        template <bool _Const = false>
        using _InnerRng = range_reference_t<_Maybe_const<_Const, _Vw>>;

```

Reply:
I'll try this, and expand `_SubRng` to `_InnerRng<_Const>` to reduce the number of similar-looking names.